src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1111 @@
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    if(fileContext->openFileMode & (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING))
+        return UA_STATUSCODE_BADINVALIDSTATE;
+
+    /* check boundaries */
+    if((size_t)length >= fileContext->file->length)
+        length = fileContext->file->length;
+
+    if((size_t)length >= (fileContext->file->length - fileContext->currentPos))
+        length = (fileContext->file->length - fileContext->currentPos);
+
+    UA_ByteString *readBuffer = UA_ByteString_new();
+    UA_ByteString_init(readBuffer);
+
+    if(length > 0) {
+        readBuffer->length = length;
+        readBuffer->data = (UA_Byte*) UA_calloc(readBuffer->length, sizeof(UA_Byte));
+        memcpy(readBuffer->data, fileContext->file->data+fileContext->currentPos, readBuffer->length);
+        fileContext->currentPos += length;
+    }
+
+    UA_Variant_setScalarCopy(output, readBuffer, &UA_TYPES[UA_TYPES_BYTESTRING]);
+    UA_ByteString_delete(readBuffer);
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+writeTrustList(UA_Server *server,
+               const UA_NodeId *sessionId, void *sessionHandle,
+               const UA_NodeId *methodId, void *methodContext,
+               const UA_NodeId *objectId, void *objectContext,
+               size_t inputSize, const UA_Variant *input,
+               size_t outputSize, UA_Variant *output) {
+    /*check for input types*/
+    if(!UA_Variant_hasScalarType(&input[0], &UA_TYPES[UA_TYPES_UINT32]) || /*FileHandle*/
+       !UA_Variant_hasScalarType(&input[1], &UA_TYPES[UA_TYPES_BYTESTRING])) /*Data*/
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+
+    UA_UInt32 fileHandle = *(UA_UInt32*)input[0].data;
+    UA_ByteString data = *(UA_ByteString*)input[1].data;
+
+    if(data.length == 0)
+        return UA_STATUSCODE_GOOD;
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    if(!certGroup->fileInfoContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_FileInfo *fileInfo = (UA_FileInfo*)certGroup->fileInfoContext;
+    UA_FileContext *fileContext = getFileContext(fileInfo, sessionId, fileHandle);
+
+    if(!fileContext)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    if(fileContext->openFileMode & UA_OPENFILEMODE_READ)
+        return UA_STATUSCODE_BADINVALIDSTATE;
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+     UA_StatusCode retval =
+         UA_decodeBinary(&data, &trustList, &UA_TYPES[UA_TYPES_TRUSTLISTDATATYPE], NULL);
+
+    if(retval != UA_STATUSCODE_GOOD)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    UA_CertificateGroup *transactionCertGroup =
+        UA_GDSTransaction_getCertificateGroup(server->transaction, certGroup);
+    if(!transactionCertGroup)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    return transactionCertGroup->setTrustList(transactionCertGroup, &trustList);
```

Comment:
As stated above, any change done through the File API must be cached elsewhere and is not part of a transaction.

A simple "Close()" will just discard those changes as will a failing "CloseAndUpdate()" call (see https://reference.opcfoundation.org/GDS/v105/docs/7.8.2.3: "If errors occur, the new TrustList is discarded. ")